Sage/carbon
# feat(dialog, sidebar, menu-full-screen): set focus to dialog container on open

### Proposed behaviour



When Dialog opens, place the focus by default on the dialog container, rather than on the first focusable element inside the dialog. This makes screen readers read out the full dialog content when it is first opened.

Also ensure that when tabbing through, the close button is focused before the content.

Also applies to DialogFullScreen, SideBar, and MenuFullScreen

closes #4975

### Current behaviour



Focus is placed on first focusable element within the dialog content. If here is a close button, even though that appears at the top of the dialog, it cannot be accessed in the tab order without going backwards (Shift+Tab) or tabbing through the entire content and cycling back to the start.

### Checklist




#### QA


### Additional context



### Testing instructions
Ensure the above-described behaviour is implemented on the following components:
- Dialog
- DialogFullScreen
- Sidebar
- MenuFullScreen
- AdvancedColorPicker
- Confirm

On all of these:
- there should be no immediately visible focus indicator on opening the dialog (or whichever component)
- when testing with a screenreader, this should make the full content of the dialog (or sidebar etc.) be announced
- pressing tab from then on goes to the first focusable element on the page (usually a close icon if this is present at the top-right corner)
- continuing to tab from there (and shift-tab likewise) should cycle through the focusable elements within the container, but never again focus the container, and also never breaking the focus trap

- Test that the above behaviour continues to work even after closing and reopening the component with no page reload, in particular make sure focus never remains outside of the component



Also test with nested dialogs to ensure the behaviour is as expected. You can test this using the following codesandbox: https://codesandbox.io/s/heuristic-jennings-26n7cn?file=/src/index.js

You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.
